docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,252 @@
+
+Other than integrating with CAPZ's `infrastructure-components.yaml`, the `capz-controller-manager` container itself could assume responsibility for installing ASO by using the Helm Go library to install the ASO chart when it starts. Installing ASO this way allows CAPZ to hook in additional logic, such as to check if ASO is already installed in the cluster or to verify ASO is healthy after it's installed, which would not be possible if a static manifest was embedded in `infrastructure-components.yaml`. However, installing ASO components from CAPZ itself would also require CAPZ to have additional RBAC permissions to create, update, and delete ASO's Deployment, Service, and other components which would not be required if the components were managed from `infrastructure-components.yaml`.
+
+For either method of installing the rest of ASO, ASO's CRDs will be installed from `infrastructure-components.yaml` to more easily facilitate `clusterctl move` and to keep the `capz-controller-manager` Pod from requiring RBAC permissions to create, update, and delete CRDs.
+
+In terms of specific options, ASO offers both [single- and multi-operator configurations](https://azure.github.io/azure-service-operator/introduction/multitenant-deployment/). The single-operator option is the default, and assumes a single ASO installation managing resources for the whole cluster. The multi-operator method allows several ASO installations to coexist, each monitoring resources in a unique set of namespaces to enable multitenancy. ASO's single-operator configuration also allows multitenancy with more tightly-scoped credentials. Overall, a single ASO installation will likely be the easier path forward since it's consistent with CAPZ's multitenancy model and doesn't involve managing ASO's CRDs and webhooks across multiple installations.
+
+The recommendation in this proposal is to install a single-operator, ["reduced permissions"](https://azure.github.io/azure-service-operator/design/adr-2023-02-helm-chart-size-limitations/#decision) configuration of ASO entirely from `infrastructure-components.yaml` to keep CAPZ from needing any new RBAC permissions to install ASO and to align with how CAPZ handles its existing dependency on AAD Pod Identity.
+
+#### Caveats
```

Comment:
I think we want to add another caveat (it seems better suited for this section than for risks, but perhaps it's more appropriate later down?) about the introduction of potentially hundreds/thousands (tens of thousands for larger clusters) of CRD resources to CAPI management clusters. I think the tl;dr is that this may impact the number of workload clusters one could reasonable maintain on a single management cluster.

Reply:
I did try to touch on this in the first paragraph of the risks section:
> Finally, integration with ASO may introduce new or worsen existing bottlenecks limiting how many CAPZ resources can be managed. Scaling bottlenecks might be Azure API rate limits or management cluster Kubernetes API server load because of the increased number of Kubernetes resources.

Does that capture what you're thinking or do you think there's something missing there?

Reply:
I think I'd add a brief not about the explicit addition of Kubernetes objects (CRD definitions resources) that can change the scaling capacity calculus for folks running large clusters and/or maintaining large numbers of clusters on a single mgmt cluster.

Reply:
Cool, expanded on that a bit more.